### PR TITLE
Phase 5: Huber Loss for Surface Pressure — Smoother Late-Training Convergence (8 parallel)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Replace the L1 surface loss with a Smooth L1 (Huber) loss using a swept delta parameter. The current pure L1 loss has constant gradient magnitude of 1 everywhere, regardless of error magnitude. This means small residual errors near the optimum receive the same gradient magnitude as large early-training errors, causing oscillation in the final epochs where EMA operates — the regime that determines final checkpoint quality.

Huber loss transitions to L2 near the origin (gradient proportional to error), providing smoother convergence in late training. This is a classic Kaggle grandmaster technique: Huber/smooth-L1 consistently outperforms pure L1 or L2 for regression, especially when the metric is MAE.

**Why this should work:**
- The "EMA + cosine absorbs modifications" pattern suggests late-training convergence quality is the binding constraint. Huber improves gradient signal exactly in this regime.
- Pure L1 has a discontinuous gradient at 0. In practice, torch gives gradient +1 or -1, causing optimizer to overshoot when errors are small. Huber smooths this.
- No architecture change needed. Fully compatible with torch.compile. Zero throughput impact.
- This has NOT been tried — loss function has always been L1 for surface pressure.

**References:** Huber (1964); Girshick "Fast R-CNN" (2015) popularized smooth L1 for box regression

## Instructions

See the first comment below for detailed implementation instructions.

## Baseline
val/loss 0.401±0.005, p_in 12.95±0.3, p_oodc 8.40±0.4, p_tan 33.8±0.5, p_re 24.7±0.2

### CODE REVIEW

**Comment by tcapelle** [comment]
## Detailed Implementation Instructions

### 1. Add config flags

```python
huber_delta: float = 0.0       # 0 = pure L1 (default), >0 = Huber with this delta for surface loss
huber_anneal: bool = False     # anneal delta from 2*huber_delta to huber_delta/4 over training
huber_pressure_only: bool = False  # apply Huber only to pressure channel, L1 for velocity
```

### 2. Replace surface loss computation

In the training loop, where surface MAE is computed, replace:

```python
# CURRENT (L1):
abs_err = (pred - y_norm).abs()
surf_mae = (abs_err * surf_mask.unsqueeze(-1)).sum() / (surf_mask.sum().clamp(min=1) * 3)

# NEW (Huber option):
if cfg.huber_delta > 0:
    delta = cfg.huber_delta
    if cfg.huber_anneal:
        progress = min(1.0, epoch / 120)
        delta = cfg.huber_delta * (2.0 - 1.75 * progress)  # 2x -> 0.25x
    
    if cfg.huber_pressure_only:
        # Huber for pressure, L1 for velocity
        p_idx = 0 if cfg.pressure_first else 2
        err = torch.zeros_like(pred)
        for c in range(3):
            if c == p_idx:
                err[:, :, c] = F.smooth_l1_loss(pred[:, :, c], y_norm[:, :, c], 
                                                  beta=delta, reduction='none')
            else:
                err[:, :, c] = (pred[:, :, c] - y_norm[:, :, c]).abs()
    else:
        err = F.smooth_l1_loss(pred, y_norm, beta=delta, reduction='none')
else:
    err = (pred - y_norm).abs()
```

**IMPORTANT:** `F.smooth_l1_loss` with `reduction='none'` returns element-wise loss. The `beta` parameter is the Huber delta.

### 3. Keep volume loss as pure L1

Only change the surface component. Volume loss should remain L1.

### 4. GPU Sweep (8 parallel)

| GPU | Config | wandb_name |
|-----|--------|------------|
| 0 | `--huber_delta 0.05` | alphonse/huber-d005 |
| 1 | `--huber_delta 0.1` | alphonse/huber-d01 |
| 2 | `--huber_delta 0.2` | alphonse/huber-d02 |
| 3 | `--huber_delta 0.5` | alphonse/huber-d05 |
| 4 | `--huber_delta 0.1 --huber_pressure_only` | alphonse/huber-d01-ponly |
| 5 | `--huber_delta 0.2 --huber_anneal` | alphonse/huber-d02-anneal |
| 6 | `--huber_delta 0.1 --huber_anneal` | alphonse/huber-d01-anneal |
| 7 | Baseline (--huber_delta 0 = pure L1) | alphonse/baseline-ctrl |

All use `--wandb_group phase5/huber-loss`.

Base command:
```bash
python train.py --agent alphonse --wandb_name "alphonse/<variant>" \
  --wandb_group phase5/huber-loss \
  --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp \
  --n_layers 3 --slice_num 96 --tandem_ramp \
  --domain_layernorm --domain_velhead --ema_decay 0.999 \
  --weight_decay 5e-5 --cosine_T_max 180 --disable_pcgrad \
  --pressure_first --pressure_deep \
  --huber_delta <D>
```

### Notes
- The key insight: in normalized Cp space, errors are O(0.1-1.0). A delta of 0.1-0.2 should affect the majority of late-training predictions.
- Log the Huber loss and L1 loss side-by-side to wandb so we can compare convergence dynamics.
- The annealed variant starts L1-like (large delta) for fast early convergence, then transitions to L2-like (small delta) for smooth late convergence.

**Comment by tcapelle** [comment]
STUDENT alphonse:

## Results

### Huber Loss Surface Pressure Sweep (8 GPUs, 159-161 epochs each)

| GPU | Config | val/loss | p_in | p_tan | p_oodc | p_re | W&B Run |
|-----|--------|----------|------|-------|--------|------|---------|
| 0 | delta=0.05 | 0.4134 | 14.2 | 34.0 | 9.6 | 25.5 | `6pwk2ax4` |
| 1 | delta=0.1 | 0.4205 | 15.1 | 34.1 | 9.9 | 25.5 | `ys3flmpg` |
| 2 | delta=0.2 | 0.4333 | 15.4 | 35.6 | 10.5 | 25.9 | `l34jo7oo` |
| 3 | delta=0.5 | 0.4335 | 15.6 | 35.3 | 11.4 | 25.9 | `8iiun2cm` |
| 4 | delta=0.1 p_only | 0.4209 | 15.0 | 34.5 | 10.1 | 25.4 | `yzdlp2j7` |
| 5 | delta=0.2 anneal | 0.4218 | 14.2 | 34.8 | 9.8 | 25.3 | `z4zhlekp` |
| 6 | delta=0.1 anneal | 0.4169 | 14.4 | 33.9 | 9.7 | 25.3 | `qbjoxzv5` |
| **7** | **Baseline (L1)** | **0.4025** | **12.8** | 34.4 | **8.3** | **24.6** | `hjqhhkw0` |

**PR Baseline**: val/loss=0.401±0.005, p_in=12.95±0.3, p_oodc=8.40±0.4, p_tan=33.8±0.5, p_re=24.7±0.2

### Implementation

Added `--huber_delta`, `--huber_anneal`, `--huber_pressure_only` config flags. When `huber_delta > 0`, surface loss uses `F.smooth_l1_loss(beta=delta)` instead of L1. Volume loss remains pure L1. Annealing ramps delta from 2x to 0.25x over 120 epochs. Also logs `train/huber_delta` and `train/surf_l1` to wandb for comparison.

### Commands

```bash
python train.py --agent alphonse --wandb_group phase5/huber-loss \
  --wandb_name "alphonse/<variant>" \
  --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp \
  --n_layers 3 --slice_num 96 --tandem_ramp \
  --domain_layernorm --domain_velhead --ema_decay 0.999 \
  --weight_decay 5e-5 --cosine_T_max 180 --disable_pcgrad \
  --pressure_first --pressure_deep --huber_delta <D>
```

### Peak memory
46-94 GB depending on config (same as baseline — Huber adds negligible overhead).

### What happened

**Huber loss is a clear negative result.** Every Huber configuration degrades all metrics:

1. **val/loss**: All Huber variants 0.413-0.434 vs baseline 0.403 (+3-8%). Monotonically worse with larger delta.

2. **p_in**: 14.2-15.6 vs baseline 12.8 (+11-22%). The strongest degradation is on in-distribution pressure — exactly the metric the Huber L2-like gradient was supposed to improve.

3. **p_oodc**: 9.6-11.4 vs baseline 8.3 (+16-37%). Generalization worsens significantly.

4. **p_tan**: 33.9-35.6 vs baseline 34.4. Interestingly, the annealed delta=0.1 variant (33.9) slightly beats baseline here — but this is within noise and doesn't compensate for other regressions.

5. **Monotonic delta effect**: Larger delta = worse. delta=0.05 is least bad (0.4134), delta=0.5 is worst (0.4335). The transition from L1→L2 near zero systematically hurts.

6. **Annealing helps slightly**: Annealed variants (GPU 5-6) are better than fixed (GPU 1-2 at same delta), confirming that L1-like early training + smoother late training is the right direction. But annealing still can't overcome the fundamental issue.

7. **Pressure-only (GPU 4)**: No advantage over full Huber (GPU 1) — the velocity channels are not the bottleneck.

### Why Huber fails

The hypothesis assumed that L1's discontinuous gradient at 0 causes oscillation in late training. But the actual dynamics are:
- With EMA (decay=0.999), gradient noise is already heavily smoothed in the weight space
- The Lion optimizer's sign-based updates are inherently robust to gradient magnitude variations
- L1's constant gradient gives strong signal on the many surface nodes with small but nonzero errors — Huber's quadratic reduction near zero weakens this signal

The model needs L1's constant pressure to push all surface predictions tighter. Huber relaxes this pressure exactly where it's most needed.

### Suggested follow-ups
- Don't pursue Huber for surface loss. L1 is optimal for MAE-evaluated CFD tasks.
- If gradient smoothing is desired, try reducing Lion's lr in the last 20 epochs instead of changing the loss function.

**Comment by tcapelle** [comment]
ADVISOR: ## Review — Huber Loss for Surface Pressure

**Result: CLOSE — Clear negative. L1 is optimal for this problem.**

Every Huber configuration degrades all metrics. Key findings:

| Delta | p_in | vs Baseline |
|-------|------|-------------|
| Baseline (L1) | 12.8 | — |
| 0.05 | 14.2 | +11% worse |
| 0.1 | 15.1 | +18% worse |
| 0.2 | 15.4 | +20% worse |
| 0.5 | 15.6 | +22% worse |

**Why L1 wins here:** Good analysis in the results. Lion optimizer's sign-based updates + EMA smoothing already handle the gradient noise that Huber was supposed to fix. L1's constant gradient magnitude pushes all predictions tighter — Huber's quadratic near-zero behavior relaxes this exactly where we need it most.

The annealing insight is useful: "L1 early + smoother late is the right direction." But Lion+EMA already achieves this implicitly.

Clean sweep, well-documented. Thanks alphonse.
